service/src/event/ordering_task.rs

```diff
@@ -175,188 +122,310 @@ impl OrderingTask {
+    /// Called when we've persisted the deliverable events to the database and can clean up our state.
+    /// Returns true if we should be retained for future processing (i.e we have more we need to discover)
+    /// and false if we can be dropped from memory.
+    fn completed_processing(&mut self) -> bool {
+        self.skip_processing = true;
+
+        for cid in self.new_deliverable.iter() {
+            if let Some(ev) = self.cid_map.get_mut(cid) {
+                match ev {
+                    StreamEvent::InitEvent(_) => {}
+                    StreamEvent::KnownDeliverable(_) => {
+                        warn!(
+                                ?ev,
+                                "Found event in deliverable queue that was already marked as deliverable."
+                            )
+                    }
+                    StreamEvent::Undelivered(meta) => {
+                        // we're delivered now
+                        *ev = StreamEvent::KnownDeliverable(meta.clone());
+                    }
+                }
+            }
         }
+        self.new_deliverable.clear();
+        self.cid_map
+            .iter()
+            .any(|(_, ev)| matches!(ev, StreamEvent::Undelivered(_)))
+    }
+
+    async fn order_events(&mut self, pool: &SqlitePool) -> Result<()> {
+        // We collect everything we can into memory and then order things.
+        // If our prev is the init event or already been delivered, we can mark ourselves as deliverable.
+        // If our prev wasn't deliverable yet, we track it and repeat (i.e. follow its prev if we don't have it)
+
+        let mut deliverable_queue = VecDeque::new();
+        let mut undelivered =
+            VecDeque::from_iter(self.cid_map.iter().filter_map(|(cid, ev)| match ev {
+                StreamEvent::Undelivered(meta) => {
+                    assert_eq!(meta.cid, *cid);
+                    Some((meta.cid, meta.prev))
+                }
+                _ => None,
+            }));
+
+        while let Some((cid, prev)) = undelivered.pop_front() {
+            if let Some(prev_event) = self.cid_map.get(&prev) {
+                match prev_event {
+                    StreamEvent::InitEvent(_) | StreamEvent::KnownDeliverable(_) => {
+                        trace!(
+                            %prev,
+                            %cid,
+                            "Found event whose prev is already in memory and IS deliverable!"
+                        );
+                        deliverable_queue.push_back(cid)
+                    }
+                    StreamEvent::Undelivered(_) => {
+                        trace!(
+                            %prev,
+                            %cid,
+                            "Found event whose prev is already in memory but NOT deliverable."
+                        );
+                        // nothing to do until it arrives on the channel
+                    }
+                }
+
+                continue;
```

Comment:
could we drop the continue and use an else block instead?  I just think it's more clear to read that way